### PR TITLE
Topological (landscape) measures for exploratory data analysis

### PR DESCRIPTION
**Pull request by ljwolf**
This adds topological measures of univariate features, `prominence` and `isolation`. 

Prominence here is measured the change in elevation between the highest low-point between two peaks. For data not on elevation, think of this in terms of ranks & features on an irregular lattice. 

Stepping down ranks (and declining in feature value),
- A "peak" is an observation that is (geographically) disconnected from all other observations that are at or above the peak's rank. Its "parent peak" is the next peak at a higher rank.
- A "key col" is an observation that, once added, connects a peak to its parent (peaks + their downslope observations) 
- A "slope" is an observation that is connected to a peak, but does not form the first connected link between a peak and its parent. 
- The "prominence" of a peak is the difference between its feature value and the feature value of its highest key col. 
- The "isolation" of a peak is the *lateral* distance between a peak and the nearest higher observation (slope, peak, or key col) 

This PR would give a few things:
1. a classifier that assigns `{peak,keycol,slope}` to observations, with the option of returning the graph built from this process.
2. an estimator that gives you the prominence of all peaks (and zero elsewhere)
3. an estimator that gives you the isolation of all peaks (and zero elsewhere)

This PR needs:
1. validation & testing that the classifier and prominence computer works correctly. 
2. a demonstration in a notebook on the example data, which isn't yet included
3. docstrings

### CODE REVIEW

**Comment by ljwolf** [comment]
@slumnitz Check for ties on the prominence measure

**Comment by ljwolf** [comment]
Working over in gdsbook/book on point patterns & Ripley's alphabet functions, I think this concept is really strongly related to concepts in point pattern analysis. For instance, the "isolation" function measures how far you have to travel from an observation before you find an increase in rank. So, maybe this can be thought of as a lattice adaptation of the K-function for marked patterns (and, obviously, lattice support)?

Prominence... not as amenable.

**Comment by ljwolf** [comment]
Ok, solved the final correctness issue with this algorithm. When a key col is found, it has to join *new* peaks to the graph. A "peak" that has been joined already is treated as one entity, so cols among its sub-peaks aren't necessarily key cols. 

For example, say we're stepping down the sorted list of input data. We find peaks/disconnected high-ranking observations
A, B, C.

We then find a *key col* between A and B, the highest ranked observation below A and B but that connects A and B. Next, we find a key col between
A and C. That peak now becomes peak "(AB)C." If we find a col between B and C, they're
*already* joined through A, so that col is just a slope on (AB)C. The predecessors go

A
A <- B
(A <- B) <- C

So, any B <- C is a connection *within* the (A <- B) <- C graph, and does not contribute to the prominence of B nor C

**Comment by ljwolf** [comment]
tests for topo.isolation, the distance from the nearest higher observation, is provided.

**Comment by ljwolf** [comment]
This is now ready to merge!

File: esda/topo.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+    either (N,) array of isolation values, or a pandas dataframe containing the full
+    tree of precedence for the isolation tree.
+    """
+    X = check_array(X, ensure_2d=False)
+    X = to_elevation(X, middle=middle).squeeze()
+    try:
+        from rtree.index import Index as SpatialIndex
+    except ImportError:
+        raise ImportError(
+            "rtree library must be installed to use " "the prominence measure"
```
Comment (jGaboardi) [inline esda/topo.py:126]: Should this simply be one string?

**Review by jGaboardi — APPROVED** [review]

**Pull request merged**